From 12167a61a85cbafb6f3591d7b432a4904e9b2dd6 Mon Sep 17 00:00:00 2001
From: Roman Stratiienko <r.stratiienko@gmail.com>
Date: Wed, 23 Jun 2021 17:10:34 +0300
Subject: [PATCH 04/21] minigbm: Add external driver support

Allow backends with custom DRM probing logic or
backends that does not rely on DRM (dma-heap, ION).

Signed-off-by: Roman Stratiienko <r.stratiienko@gmail.com>
Change-Id: I7bcaf10205ca051eb109d6e220b8a2af38267442
---
 cros_gralloc/cros_gralloc_driver.cc | 13 ++++++++++++-
 drv.c                               | 10 ++++++++++
 2 files changed, 22 insertions(+), 1 deletion(-)

diff --git a/cros_gralloc/cros_gralloc_driver.cc b/cros_gralloc/cros_gralloc_driver.cc
index 7bc1620..3536306 100644
--- a/cros_gralloc/cros_gralloc_driver.cc
+++ b/cros_gralloc/cros_gralloc_driver.cc
@@ -89,6 +89,7 @@ cros_gralloc_driver *cros_gralloc_driver::get_instance()
 	return &s_instance;
 }
 
+#ifndef DRV_EXTERNAL
 static struct driver *init_try_node(int idx, char const *str)
 {
 	int fd;
@@ -146,11 +147,21 @@ static struct driver *init_try_nodes()
 	return nullptr;
 }
 
+#else
+
+static struct driver *init_try_nodes()
+{
+	return drv_create(-1);
+}
+
+#endif
+
 static void drv_destroy_and_close(struct driver *drv)
 {
 	int fd = drv_get_fd(drv);
 	drv_destroy(drv);
-	close(fd);
+	if (fd != -1)
+		close(fd);
 }
 
 cros_gralloc_driver::cros_gralloc_driver() : drv_(init_try_nodes(), drv_destroy_and_close)
diff --git a/drv.c b/drv.c
index 52abe26..59996bf 100644
--- a/drv.c
+++ b/drv.c
@@ -24,6 +24,15 @@
 #include "drv_priv.h"
 #include "util.h"
 
+#ifdef DRV_EXTERNAL
+extern struct backend *init_external_backend();
+
+static const struct backend *drv_get_backend(int fd)
+{
+	return init_external_backend();
+}
+#else
+
 #ifdef DRV_AMDGPU
 extern const struct backend backend_amdgpu;
 #endif
@@ -102,6 +111,7 @@ static const struct backend *drv_get_backend(int fd)
 	drmFreeVersion(drm_version);
 	return NULL;
 }
+#endif
 
 struct driver *drv_create(int fd)
 {
-- 
2.39.2

